cmd/gpu_plugin/rm/gpu_plugin_resource_manager_test.go

```diff
@@ -0,0 +1,228 @@
+type mockClient struct {
+	fake.Clientset
+	mockCoreV1
+}
+
+type mockCoreV1 struct {
+	fakecorev1.FakeCoreV1
+	mockPods
+}
+
+type mockPods struct {
+	fakecorev1.FakePods
+	tc testCase
+}
+
+func (m *mockClient) CoreV1() corev1.CoreV1Interface {
+	return m
+}
+
+func (m *mockCoreV1) Pods(namespace string) corev1.PodInterface {
+	return m
+}
+
+func (m *mockPods) List(ctx context.Context, opts metav1.ListOptions) (*v1.PodList, error) {
+	return &v1.PodList{
+		Items: m.tc.pods,
+	}, nil
+}
+
+type mockPodresourcesWrapper struct {
+	client *grpc.ClientConn
+	tc     testCase
+}
+
+func (w *mockPodresourcesWrapper) List(ctx context.Context,
+	in *podresourcesv1.ListPodResourcesRequest,
+	opts ...grpc.CallOption) (*podresourcesv1.ListPodResourcesResponse, error) {
+	resp := podresourcesv1.ListPodResourcesResponse{}
+	for _, pod := range w.tc.pods {
+		resp.PodResources = append(resp.PodResources, &podresourcesv1.PodResources{
+			Name: pod.ObjectMeta.Name, Containers: []*podresourcesv1.ContainerResources{{}},
+		})
+	}
+	return &resp, nil
+}
+func (w *mockPodresourcesWrapper) GetAllocatableResources(ctx context.Context,
+	in *podresourcesv1.AllocatableResourcesRequest,
+	opts ...grpc.CallOption) (*podresourcesv1.AllocatableResourcesResponse, error) {
+	return nil, nil
+}
+
+func (w *mockPodresourcesWrapper) GetV1Client(socket string,
+	connectionTimeout time.Duration,
+	maxMsgSize int) (podresourcesv1.PodResourcesListerClient, *grpc.ClientConn, error) {
+	return w, w.client, nil
+}
+
+func newMockResourceManager(tc testCase) *ResourceManager {
+	client, err := grpc.Dial("", grpc.WithInsecure())
+	if err != nil {
+		os.Exit(1)
+	}
+
+	mc := &mockClient{}
+	mc.mockCoreV1.mockPods.tc = tc
+	rm := ResourceManager{
+		clientset:    mc,
+		nodeName:     "TestNode",
+		podresources: &mockPodresourcesWrapper{client: client, tc: tc},
+	}
+
+	deviceInfoMap := NewDeviceInfoMap()
+	deviceInfoMap["card0-0"] = NewDeviceInfo([]v1beta1.DeviceSpec{{}},
+		[]v1beta1.Mount{{}}, map[string]string{"more": "coverage"})
+	deviceInfoMap["card1-0"] = NewDeviceInfo([]v1beta1.DeviceSpec{{}}, nil, nil)
+	rm.SetDevInfos(deviceInfoMap)
+
+	return &rm
+}
+
+func TestConstruction(t *testing.T) {
+	// normal clientset is unavailable inside the unit tests
+	_, err := NewResourceManager()
+
+	if err == nil {
+		t.Errorf("unexpected success")
+	}
+}
+
+func TestInputChecks(t *testing.T) {
+	rm := newMockResourceManager(testCase{})
+
+	// test wrong amount of container requests
+	_, err := rm.ReallocateWithFractionalResources(&v1beta1.AllocateRequest{})
+	if err == nil {
+		t.Errorf("unexpected success")
+	}
+
+	// test monitor request
+	_, err = rm.ReallocateWithFractionalResources(&v1beta1.AllocateRequest{
+		ContainerRequests: []*v1beta1.ContainerAllocateRequest{
+			{
+				DevicesIDs: []string{
+					monitorID,
+				},
+			},
+		},
+	})
+	if err == nil {
+		t.Errorf("unexpected success")
+	}
+}
+
+func TestE2E(t *testing.T) {
```

Comment:
This function and `TestInputChecks` together seem to be a test for the `ReallocateWithFractionalResources() unit`. They should be merged into `TestRealocateWithFractionalResources()`.

Then tests for other units would be nice to have: this way it's easier to notice deficiencies in their design.

For example `TestConstruction()` seems to be a test for `NewResourceManager()` (hence should be named `TestNewResourceManager()`). But it fails to test the case when real resources are used in `rest.InClusterConfig()`. In this particular case I'm Ok not to test this branch, but theoretically this behavior of `NewResourceManager()` could be parameterized in `main()`.

Reply:
Done (merging part)